core/src/main/java/com/linecorp/armeria/client/retry/AbstractRetryingClient.java

```diff
@@ -69,69 +68,37 @@
     private static final AttributeKey<State> STATE =
             AttributeKey.valueOf(AbstractRetryingClient.class, "STATE");
 
-    @Nullable
-    private final RetryRule retryRule;
-
-    @Nullable
-    private final RetryRule fromRetryRuleWithContent;
-
-    @Nullable
-    private final RetryRuleWithContent<O> retryRuleWithContent;
-
-    private final int maxTotalAttempts;
-    private final long responseTimeoutMillisForEachAttempt;
-
-    /**
-     * Creates a new instance that decorates the specified {@link Client}.
-     */
-    AbstractRetryingClient(Client<I, O> delegate, RetryRule retryRule,
-                           int maxTotalAttempts, long responseTimeoutMillisForEachAttempt) {
-        this(delegate, requireNonNull(retryRule, "retryRule"), null,
-             maxTotalAttempts, responseTimeoutMillisForEachAttempt);
-    }
+    private final RetryConfigMapping<O> mapping;
+    private final RetryConfig retryConfig;
 
     /**
      * Creates a new instance that decorates the specified {@link Client}.
      */
-    AbstractRetryingClient(Client<I, O> delegate,
-                           RetryRuleWithContent<O> retryRuleWithContent,
-                           int maxTotalAttempts, long responseTimeoutMillisForEachAttempt) {
-        this(delegate, null, requireNonNull(retryRuleWithContent, "retryRuleWithContent"),
-             maxTotalAttempts, responseTimeoutMillisForEachAttempt);
-    }
-
-    /**
-     * Creates a new instance that decorates the specified {@link Client}.
-     */
-    private AbstractRetryingClient(Client<I, O> delegate, @Nullable RetryRule retryRule,
-                                   @Nullable RetryRuleWithContent<O> retryRuleWithContent,
-                                   int maxTotalAttempts, long responseTimeoutMillisForEachAttempt) {
+    AbstractRetryingClient(
+            Client<I, O> delegate, RetryConfigMapping<O> mapping, @Nullable RetryConfig<O> retryConfig) {
```

Comment:
There are protected getters in this class: retryRule() and retryRuleWithContent(), which need the retryConfig to pull the rule from.

mapping is never null. Even if it wasn't explicitly set and a single retryConfig was set instead, mapping is set as a function that always returns that config. mapping is used everywhere to get rule and config. However, the getters mentioned above can still be useful for people not setting a mapping, and they need to know for sure whether a retryConfig was explicitly set.

Reply:
> the getters mentioned above can still be useful for people not setting a mapping, and they need to know for sure whether a retryConfig was explicitly set.

I was sort of thinking to store the `retryConfig` to `ClientRequestContext` in the `execute()` method so that we can call `ClientReqesutContext.current()` and retrieve the `retryConfig` in `retryRule()` method.
But, I'm also fine as it is now. 😄